### PR TITLE
Propagate chunk_layout to Python meep.Simulation object

### LINKED ISSUE #1648: sim.chunk_layout is not accessible in Python if not specified at class instantiation time
If a `meep.Simulation` object is created without specifying `chunk_layout`, then the [default logic](https://github.com/NanoComp/meep/blob/master/src/structure.cpp#L109) for choosing chunk divisions is invoked when `sim.init_sim()` is called. However, the `chunk_layout` attribute on the Python `Simulation` object is not updated to reflect this change and will still read as `None`.

When converting from a Python `meep.BinaryPartition` object, [`py_bp_to_bp()`](https://github.com/NanoComp/meep/blob/master/python/typemap_utils.cpp#L1045) is called to convert the Python object into a C++ object. It would be useful to have the inverse operation present with a `bp_to_py_bp()` function, perhaps as a class construction method to live on `meep.BinaryPartition`. Then in `Simulation._init_structure()` at [line 1709](https://github.com/NanoComp/meep/blob/master/python/simulation.py#L1708), this function could be called to update the `Simulation.chunk_layout` attribute:

```python
if self.chunk_layout is None:
  self.chunk_layout = mp.BinaryPartition.from_cpp_object(...)
```

### PR DESCRIPTION
**Pull request by bencbartlett**
If a `meep.Simulation` object is created without specifying `chunk_layout`, then the default logic for choosing chunk partitionings is invoked when `sim.init_sim()` is called. However, the `chunk_layout` attribute on the Python `meep.Simulation` object previously was not updated and would still read as `None`. This pull request adds a `bp_to_py_bp()` function to convert the C++ object into a Python `meep.BinaryPartition` object and updates the `sim.chunk_layout` attribute (and for consistency, `sim.num_chunks`), in `sim._init_structure()`. (Fixes #1648)

### CODE REVIEW

**Comment by oskooi** [comment]
I think the failing test `test_chunk_layout.py` is due to a bug in the way the process IDs of the leaves are *always* assigned a value of -1:

https://github.com/NanoComp/meep/blob/89c93498ead04290b42acbb35b2c09c9813e04eb/src/structure.cpp#L69

The reason this bug is a little hard to track down is because if the user does not pass in a `chunk_layout` parameter to the `Simulation` object and Meep therefore falls back on using its internal cell-partition scheme, the `proc` of each chunk is set correctly but the corresponding `proc_id` of the leaves from the binary tree is never updated to this same value:

https://github.com/NanoComp/meep/blob/89c93498ead04290b42acbb35b2c09c9813e04eb/src/structure.cpp#L133-L141

What should happen is the `proc_id` of each leaf/chunk should be set to `proc` *before* returning the `binary_partition` object `bp` back to Python.

File: .gitignore [hunk 1]
```diff
@@ -14,6 +14,7 @@
 *.pyc
 _build
 .vscode
+.idea
```
Comment (oskooi) [inline .gitignore:17]: Is this necessary?

File: src/structure.cpp [hunk 2]
```diff
@@ -958,6 +959,7 @@ double structure_chunk::max_eps() const {
 }
 
 bool structure::equal_layout(const structure &s) const {
+  std::cout << "equal_layout num_chunks " << num_chunks << " s.num_chunks " << s.num_chunks << "\n";
```
Comment (oskooi) [inline src/structure.cpp:962]: Is this something that was left over from debugging and therefore should be removed from this PR? 

In general, for printing to standard output in Meep, use the built-in function `master_printf("equal_layout %d s.num_chunks %d\n",num_chunks,s.num_chunks);` which will only print from the master process in a parallel simulation.

File: src/structure.cpp [hunk 3]
```diff
@@ -15,6 +15,7 @@
 %  Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
+#include <iostream>
```
Comment (oskooi) [inline src/structure.cpp:18]: This should probably be removed (see comment below).

File: src/structure.cpp (hunk 2 above)
Comment (bencbartlett) [reply src/structure.cpp]: Removed this line

File: src/structure.cpp (hunk 3 above)
Comment (bencbartlett) [reply src/structure.cpp]: Done

File: .gitignore (hunk 1 above)
Comment (bencbartlett) [reply .gitignore:17]: I used CLion to make these changes locally. JetBrains IDEs (CLion, IntelliJ) are pretty popular so it's probably worth including in the .gitignore file.

**Review by oskooi — APPROVED** [review]

**Pull request merged**